geoapps/octree_creation/driver.py

```diff
@@ -77,46 +80,174 @@ def octree_from_params(params: OctreeParams):
+                treemesh = OctreeDriver.refine_tree_from_points(
+                    treemesh,
+                    refinement_object,
+                    levels,
+                )
+
+            else:
+                raise NotImplementedError(
+                    f"Refinement type {value['type']} is not implemented."
                 )
 
         print("Finalizing . . .")
         treemesh.finalize()
-
         octree = treemesh_2_octree(params.geoh5, treemesh, name=params.ga_group_name)
 
         return octree
 
     @staticmethod
     def refine_tree_from_curve(
-        treemesh: TreeMesh, curve: Curve, levels: list[int], finalize=False
+        treemesh: TreeMesh, curve: Curve, levels: list[int] | np.ndarray, finalize=False
     ) -> TreeMesh:
         """
-        Refine a tree mesh along the segments of a curve.
+        Refine a tree mesh along the segments of a curve densified by the
+        mesh cell size.
 
         :param treemesh: Tree mesh to refine.
         :param curve: Curve object to use for refinement.
         :param levels: Number of cells requested at each refinement level.
             Defined in reversed order from highest octree to lowest.
         """
+        if not isinstance(curve, Curve):
+            raise TypeError("Refinement object must be a Curve.")
+
+        if isinstance(levels, list):
+            levels = np.array(levels)
+
         locs = densify_curve(curve, treemesh.h[0][0])
-        distance = 0
+        treemesh = OctreeDriver.refine_tree_from_points(
+            treemesh, locs, levels, finalize=False
+        )
+
+        if finalize:
+            treemesh.finalize()
+
+        return treemesh
+
+    @staticmethod
+    def refine_tree_from_points(
+        treemesh: TreeMesh,
+        points: ObjectBase | np.ndarray,
+        levels: list[int] | np.ndarray,
+        finalize=False,
+    ) -> TreeMesh:
+        """
+        Refine a tree mesh along the vertices of an object.
+
+        :param treemesh: Tree mesh to refine.
+        :param points: Object to use for refinement.
+        :param levels: Number of cells requested at each refinement level.
+            Defined in reversed order from highest octree to lowest.
+        :param finalize: Finalize the tree mesh after refinement.
+
+        :return: Refined tree mesh.
+        """
+        if isinstance(points, ObjectBase):
+            locs = get_locations(points.workspace, points)
+        else:
+            locs = points
+
+        if locs is None:
+            raise ValueError("Could not find locations for refinement.")
 
+        if isinstance(levels, list):
+            levels = np.array(levels)
+
+        distance = 0
         for ii, n_cells in enumerate(levels):
             distance += n_cells * treemesh.h[0][0] * 2**ii
 
             treemesh.refine_ball(
-                locs, distance, treemesh.max_level - ii, finalize=finalize
+                locs, distance, treemesh.max_level - ii, finalize=False
```

Comment:
why not just pass the finalize input argument here?

Reply:
Because we want to get through the loop for the different levels first. Otherwise it could finalize the mesh after the level,, and try to finalize again on the second iteration (which is bad).